src/main/java/com/spotify/futures/CompletableFutures.java

```diff
@@ -156,6 +160,38 @@ private CompletableFutures() {
     return collectingAndThen(toList(), CompletableFutures::allAsList);
   }
 
+  /**
+   * Apply an asynchronous operation to every entity in a stream, and collect all replies into a
+   * map.
+   *
+   * <p> Usage:
+   *
+   * <pre>{@code
+   *   collection.stream()
+   *      .collect(joinMap(this::someAsyncFunc))
+   *      .thenApply(this::consumeMap)
+   * }</pre>
+   *
+   * <p> The generated {@link CompletableFuture} will complete to a map from entity to the result
+   * of the supplied asynchronous opoertaion.  Similar to
+   * {@link CompletableFuture#allOf(CompletableFuture[])}, if any of the map operations complete
+   * exceptionally, then the returned CompletableFuture also does so, with a
+   * {@link CompletionException} holding this exception as its cause.
+   *
+   * @param <K> the key type in the output map
+   * @param <V> the value type in the output map
+   * @return a new {@link CompletableFuture} according to the rules outlined in the method
+   * description
+   * @since 0.3.3
+   */
+  public static <K, V> Collector<K, ?, CompletableFuture<Map<K, V>>> joinMap(
+      final Function<? super K, CompletionStage<V>> func) {
+    return collectingAndThen(toList(), keys -> keys.stream()
```

Comment:
Changing indentation here would make the code a little more readable - I had a hard time figuring out how `keys` could be in scope at first. It looked like the `.map`, etc., were being applied to the result of `collectingAndThen`.

Reply:
Agreed, updated!